clippy_lints/src/methods/string_as_str.rs

```diff
@@ -0,0 +1,60 @@
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::is_type_diagnostic_item};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::LateContext;
+
+use super::STRING_AS_STR;
+
+pub(super) fn check(cx: &LateContext<'_>, expr: &Expr<'_>) {
+    if let ExprKind::MethodCall(path, sub_expr, ..) = &expr.kind {
+        if is_string(cx, sub_expr) && !is_within_borrowed_call(cx, sub_expr) {
+            let var_name = if let ExprKind::Path(QPath::Resolved(_, path)) = sub_expr.kind {
+                path.segments.first().map(|name| name.ident.name.as_str())
+            } else {
+                None
+            };
+
+            span_lint_and_sugg(
+                cx,
+                STRING_AS_STR,
+                expr.span,
+                &format!("used `{}` for reference", path.ident.name),
+                "replace it with",
+                format!("&{}", var_name.unwrap_or("")),
+                Applicability::MaybeIncorrect,
+            );
+        }
+    }
+}
+
+fn is_string(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    is_type_diagnostic_item(
+        cx,
+        cx.typeck_results().expr_ty(expr).peel_refs(),
+        rustc_span::sym::String,
+    )
+}
+
+fn is_within_borrowed_call(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
```

Comment:
I'm not sure what this function is supposed to do. It looks like it checks for really specific things in the tests you've written.

Reply:
Don't know how to not lint here https://github.com/rust-lang/rust-clippy/blob/50f192f86a12d4b39d8f7f061fbd8e0ba8007932/clippy_lints/src/missing_enforced_import_rename.rs#L77

Reply:
Why can't this be linted?

Reply:
Because using `&snip` there are different type on those two branch, one is &str and other &String, don't know if we can change the code in other way

Reply:
Exactly... Don't sure next step to complete this lint :( 🤔